src/cargo/util/command_prelude.rs

```diff
@@ -1174,6 +1178,122 @@ fn get_target_triples_from_rustc() -> CargoResult<Vec<clap_complete::CompletionC
+                    .push(package);
+                map
+            });
+
+    package_map
+        .into_iter()
+        .flat_map(|(name, packages)| {
+            // For unique package name
+            if packages.len() == 1 {
+                return vec![
+                    clap_complete::CompletionCandidate::new(name.to_string()).help(
+                        packages[0]
+                            .manifest()
+                            .metadata()
+                            .description
+                            .to_owned()
+                            .map(From::from),
+                    ),
+                ];
+            }
+
+            let version_map = HashMap::<Version, Vec<Package>>::new();
+            let version_map = packages.into_iter().fold(version_map, |mut map, package| {
+                map.entry(package.version().to_owned())
+                    .or_insert_with(Vec::new)
+                    .push(package);
+                map
+            });
+
+            version_map
+                .into_iter()
+                .flat_map(|(version, packages)| {
+                    // For package name with duplicates but unique version
+                    if packages.len() == 1 {
+                        return vec![clap_complete::CompletionCandidate::new(format!(
+                            "{}@{}",
+                            name, version
+                        ))
+                        .help(
+                            packages[0]
+                                .manifest()
+                                .metadata()
+                                .description
+                                .to_owned()
+                                .map(From::from),
+                        )];
+                    }
+
+                    // For package name with duplicates and duplicate version
+                    packages
+                        .into_iter()
+                        .map(|package| {
+                            clap_complete::CompletionCandidate::new(format!(
+                                "{}",
+                                package.package_id().to_spec()
+                            ))
+                            .help(
+                                package
+                                    .manifest()
+                                    .metadata()
+                                    .description
+                                    .to_owned()
+                                    .map(From::from),
+                            )
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>()
+}
+
+fn get_packages() -> CargoResult<Vec<Package>> {
+    let cwd = std::env::current_dir()?;
+    let mut gctx = GlobalContext::new(shell::Shell::new(), cwd.clone(), cargo_home_with_cwd(&cwd)?);
+    gctx.configure(0, true, None, false, true, false, &None, &[], &[])?;
```

Comment:
Could you split these out into variables to clarify their names?

We'll likely need this in all of the completions that create a context, so maybe we should provide a helper for this?  Feels too specialized and accident prone to live on `GlobalContext` but unsure.

Reply:
> We'll likely need this in all of the completions that create a context, so maybe we should provide a helper for this?

I will refactor other completers used `gctx` in another pr.

> Feels too specialized and accident prone to live on GlobalContext but unsure.

For now, I make this function a non-pub function for now. I think putting it in the impl of `GlobalContext` is too specialized.